index.rst

```diff
@@ -48,14 +48,126 @@
+This SQLite database would still refer to the datasets in their original datastore location.
+The pipeline would then be executed using the local registry and would interact as before with the cloud object store.
+When the processing completes the newly-created records would be exported from the local registry and imported to the cloud registry.
+This would use the automatic transfer mode since the files are already present in the object store in the correct place.
+
+If quanta are chained, with the outputs of one ``PipelineTask`` feeding directly to the inputs of the next within the same job, the datastore will be configured to do local caching and write the file locally as well as to the object store so that the next one would read the file directly.
+
+Doing this would have the advantage that the object store writes are no longer linked to a long-lived transaction in the shared database.
+On the other hand, if the pipeline processing fails in some way it is likely that there will be files in the object store that are not present in the cloud registry.
+These files can be deleted in a clean up job or they can be ignored, allowing them to be over written.
+Regardless, if the node is pre-empted during processing after files have been written, the orphaned files will be present and there will be no associated registry entries so attempting to clean-up does not gain us anything other than saving storage space if that run is never used again.
+The execution system is already capable of over-writing an existing file if it is not known to registry and datastore ensures that the file name is unique for a specific dataId and run combination.
+
+If pre-emption occurs during the final registry import it is assumed that there is a single transaction that will be completely rolled back.
+The end result would be that the entire processing would have to be redone.
+If the pre-emption occurs in the short time between completion of the registry import and the completion of the job, we should ensure that the ``pipetask`` command is configured to complete without action when the job is restarted.
+
+Externalized
+------------
+
+The most flexible approach is for the workflow generator to insert extra jobs into the workflow that handle registry merging and registry syncing.
+
+A new job would be inserted before the initial pipeline jobs in the graph that downloads the initial registry state from the cloud registry and stores it in a SQLite file.
+This SQLite file would then be provided as an input to the first job.
+The pipeline job would use that SQLite registry and a GCS-with-local-cache datastore as described in the previous section.
+On completion the output collected by the workflow execution system would be an updated SQLite registry.
+If the next pipeline job only takes files from a single input job this SQLite file would be passed directly to the next pipeline job.
```

Comment:
What if the next pipeline job requires outputs from only a single job but additional raw or calibration inputs?  Does that require a merge job?

Reply:
The sqlite file could already have all the calibration entries in it couldn't it? The question is where in the graph we add them. Adding all calibrations right at the front would work but you might end up with a sqlite file that is too large.